File: sort.go
package main

import (
	"fmt"
	"log/slog"
	"math"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/disgoorg/disgo/discord"
	"github.com/disgoorg/disgo/events"
)

type SortableItem struct {
	Item  *Item
	Value int
}

const ItemsPerPage = 10

func CommandSort(e *events.ApplicationCommandInteractionCreate) {
	statType := e.SlashCommandInteractionData().String("stat")
	itemType := e.SlashCommandInteractionData().String("type") // optional

	var filteredItems []*Item
	for _, item := range ItemsData {
		if item.Deleted || item.Name == "None" || len(item.StatsPerLevel) == 0 {
			continue
		}

		// apply filters if any
		if itemType != "" && item.MainType != itemType {
			continue
		}

		if hasStatAtLevel140(&item, statType) {
			filteredItems = append(filteredItems, &item)
		}
	}

	if len(filteredItems) == 0 {
		err := e.CreateMessage(
			discord.NewMessageCreateBuilder().
				SetContent(fmt.Sprintf("No items found with %s stats%s.", statType, getTypeFilter(itemType))).
				SetEphemeral(true).
				Build(),
		)
		if err != nil {
			slog.Error("error sending no items message", slog.Any("err", err))
		}
		return
	}

	// create sortable items with their stat values at level 140
	var sortableItems []SortableItem
	for _, item := range filteredItems {
		value := getStatValueAtLevel140(item, statType)
		if value > 0 {
			sortableItems = append(sortableItems, SortableItem{
				Item:  item,
				Value: value,
			})
		}
	}

	sort.Slice(sortableItems, func(i, j int) bool {
		return sortableItems[i].Value > sortableItems[j].Value
	})

	totalPages := int(math.Ceil(float64(len(sortableItems)) / float64(ItemsPerPage)))
	currentPage := 1

	embed := buildSortEmbed(sortableItems, statType, itemType, currentPage, totalPages)
	components := buildPaginationComponents(currentPage, totalPages, statType, itemType)

	err := e.CreateMessage(
		discord.NewMessageCreateBuilder().
			AddEmbeds(embed).
			AddActionRow(components...).
			Build(),
	)

	if err != nil {
		slog.Error("error sending sort message", slog.Any("err", err))
	}
}

func hasStatAtLevel140(item *Item, statType string) bool {
	for _, stats := range item.StatsPerLevel {
		if stats.Level == 140 {
			switch statType {
			case "power":
				return stats.Power > 0
			case "agility":
				return stats.Agility > 0
			case "attackspeed":
				return stats.AttackSpeed > 0
			case "defense":
				return stats.Defense > 0
			case "attacksize":
				return stats.AttackSize > 0
			case "intensity":
				return stats.Intensity > 0
			case "regeneration":
				return stats.Regeneration > 0
			case "resistance":
				return stats.Resistance > 0
			case "armorpiercing":
				return stats.Piercing > 0
			}
		}
	}
	return false
}

func getStatValueAtLevel140(item *Item, statType string) int {
	for _, stats := range item.StatsPerLevel {
		if stats.Level == 140 {
			switch statType {
			case "power":
				return stats.Power
			case "agility":
				return stats.Agility
			case "attackspeed":
				return stats.AttackSpeed
			case "defense":
				return stats.Defense
			case "attacksize":
				return stats.AttackSize
			case "intensity":
				return stats.Intensity
			case "regeneration":
				return stats.Regeneration
			case "resistance":
				return stats.Resistance
			case "armorpiercing":
				return stats.Piercing

			}
		}
	}
	return 0
}

func getTypeFilter(itemType string) string {
	if itemType != "" {
		return fmt.Sprintf(" for %s items", strings.ToLower(itemType))
	}
	return ""
}

func getStatDisplayName(statType string) string {
	switch statType {
	case "power":
		return "Power"
	case "agility":
		return "Agility"
	case "attackspeed":
		return "Attack Speed"
	case "defense":
		return "Defense"
	case "attacksize":
		return "Attack Size"
	case "intensity":
		return "Intensity"
	case "regeneration":
		return "Regeneration"
	case "resistance":
		return "Resistance"
	case "armorpiercing":
		return "Armor Piercing"

	default:
		// capitalize first letter manually
		if len(statType) > 0 {
			return strings.ToUpper(statType[:1]) + statType[1:]
		}
		return statType
	}
}

func buildSortEmbed(sortableItems []SortableItem, statType, itemType string, currentPage, totalPages int) discord.Embed {
	startIndex := (currentPage - 1) * ItemsPerPage
	endIndex := startIndex + ItemsPerPage
	if endIndex > len(sortableItems) {
		endIndex = len(sortableItems)
	}

	pageItems := sortableItems[startIndex:endIndex]

	var description strings.Builder
	description.WriteString(fmt.Sprintf("Items sorted by %s at level 140%s\n\n", getStatDisplayName(statType), getTypeFilter(itemType)))

	for i, sortableItem := range pageItems {
		rank := startIndex + i + 1
		item := sortableItem.Item

		description.WriteString(fmt.Sprintf("**%d.** %s\n", rank, item.Name))
		description.WriteString(fmt.Sprintf("   %s: **%d** | %s", getStatDisplayName(statType), sortableItem.Value, item.Rarity))

		if item.SubType != "" {
			description.WriteString(fmt.Sprintf(" | %s", item.SubType))
		}

		description.WriteString("\n\n")
	}

	embed := discord.NewEmbedBuilder().
		SetTitle(fmt.Sprintf("Top Items by %s", getStatDisplayName(statType))).
		SetDescription(description.String()).
		SetColor(DefaultColor).
		SetFooter(fmt.Sprintf("Page %d/%d • %s", currentPage, totalPages, EmbedFooter), "").
		SetTimestamp(time.Now())

	return embed.Build()
}

func buildPaginationComponents(currentPage, totalPages int, statType, itemType string) []discord.InteractiveComponent {
	var components []discord.InteractiveComponent

	prevButton := discord.NewSecondaryButton("◀ Prev", fmt.Sprintf("sort_prev_%s_%s_%d", statType, itemType, currentPage))
	if currentPage <= 1 {
		prevButton = prevButton.AsDisabled()
	}
	components = append(components, prevButton)

	pageButton := discord.NewSecondaryButton(fmt.Sprintf("Page %d/%d", currentPage, totalPages), "sort_page_indicator").AsDisabled()
	components = append(components, pageButton)

	nextButton := discord.NewSecondaryButton("Next ▶", fmt.Sprintf("sort_next_%s_%s_%d", statType, itemType, currentPage))
	if currentPage >= totalPages {
		nextButton = nextButton.AsDisabled()
	}
	components = append(components, nextButton)

	return components
}

func handleSortPagination(e *events.ComponentInteractionCreate) {
	customID := e.ButtonInteractionData().CustomID()

	parts := strings.Split(customID, "_")
	if len(parts) < 5 {
		slog.Error("invalid sort pagination custom ID", "customID", customID)
		return
	}

	action := parts[1]   // "prev" or "next"
	statType := parts[2] // stat type
	itemType := parts[3] // item type (may be empty)
	currentPageStr := parts[4]

	currentPage, err := strconv.Atoi(currentPageStr)
	if err != nil {
		slog.Error("error parsing current page", slog.Any("err", err))
		return
	}

	// calculate new page
	var newPage int
	switch action {
	case "prev":
		newPage = currentPage - 1
	case "next":
		newPage = currentPage + 1
	default:
		return
	}

	// regenerate the sorted items (this could be optimized by caching)
	var filteredItems []*Item
	for _, item := range ItemsData {
		if item.Deleted || item.Name == "None" || len(item.StatsPerLevel) == 0 {
			continue
		}

		if itemType != "" && item.MainType != itemType {
			continue
		}

		if hasStatAtLevel140(&item, statType) {
			filteredItems = append(filteredItems, &item)
		}
	}

	var sortableItems []SortableItem
	for _, item := range filteredItems {
		value := getStatValueAtLevel140(item, statType)
		if value > 0 {
			sortableItems = append(sortableItems, SortableItem{
				Item:  item,
				Value: value,
			})
		}
	}

	sort.Slice(sortableItems, func(i, j int) bool {
		return sortableItems[i].Value > sortableItems[j].Value
	})

	totalPages := int(math.Ceil(float64(len(sortableItems)) / float64(ItemsPerPage)))

	// validate new page
	if newPage < 1 {
		newPage = 1
	}
	if newPage > totalPages {
		newPage = totalPages
	}

	// build new embed and components
	embed := buildSortEmbed(sortableItems, statType, itemType, newPage, totalPages)
	components := buildPaginationComponents(newPage, totalPages, statType, itemType)

	err = e.UpdateMessage(
		discord.NewMessageUpdateBuilder().
			AddEmbeds(embed).
			AddActionRow(components...).
			Build(),
	)

	if err != nil {
		slog.Error("error updating sort pagination", slog.Any("err", err))
	}
}

File: src/commands/sort.rs
use crate::utils::{filter_and_sort_items, get_stat_display_name, get_type_filter, SortableItem};
use crate::{Context, Error, DEFAULT_COLOR, EMBED_FOOTER};
use poise::serenity_prelude as serenity;
use serenity::builder::CreateEmbed;

/// Sort and display items by specific stats
#[poise::command(slash_command)]
pub async fn sort(
    ctx: Context<'_>,
    #[description = "The stat to sort by"]
    #[autocomplete = "stat_autocomplete"]
    stat: String,
    #[description = "Filter by item type (optional)"]
    #[autocomplete = "item_type_autocomplete"]
    item_type: Option<String>,
) -> Result<(), Error> {
    let data = ctx.data();

    // Validate stat type
    let valid_stats = [
        "power",
        "agility",
        "attackspeed",
        "defense",
        "attacksize",
        "intensity",
        "regeneration",
        "resistance",
        "armorpiercing",
    ];

    if !valid_stats.contains(&stat.as_str()) {
        ctx.say(format!(
            "Invalid stat type! Valid stats are: {}",
            valid_stats.join(", ")
        ))
        .await?;
        return Ok(());
    }

    // Filter and sort items
    let sortable_items = filter_and_sort_items(data, &stat, item_type.as_deref()).await;

    if sortable_items.is_empty() {
        let type_filter = get_type_filter(item_type.as_deref());
        ctx.say(format!(
            "No items found with {} stats{}.",
            stat, type_filter
        ))
        .await?;
        return Ok(());
    }

    let total_pages = ((sortable_items.len() as f64) / 10.0).ceil() as usize;
    let current_page = 1;

    let embed = build_sort_embed(
        &sortable_items,
        &stat,
        item_type.as_deref(),
        current_page,
        total_pages,
    );
    let components =
        build_pagination_components(current_page, total_pages, &stat, item_type.as_deref());

    ctx.send(
        poise::CreateReply::default()
            .embed(embed)
            .components(components),
    )
    .await?;

    Ok(())
}

// Autocomplete functions for sort command
pub async fn stat_autocomplete(_ctx: Context<'_>, partial: &str) -> Vec<String> {
    let stats = [
        ("power", "Power"),
        ("agility", "Agility"),
        ("attackspeed", "Attack Speed"),
        ("defense", "Defense"),
        ("attacksize", "Attack Size"),
        ("intensity", "Intensity"),
        ("regeneration", "Regeneration"),
        ("resistance", "Resistance"),
        ("armorpiercing", "Armor Piercing"),
    ];

    stats
        .iter()
        .filter(|(value, name)| {
            value.contains(&partial.to_lowercase())
                || name.to_lowercase().contains(&partial.to_lowercase())
        })
        .map(|(value, _name)| (*value).to_string())
        .collect()
}

pub async fn item_type_autocomplete(ctx: Context<'_>, partial: &str) -> Vec<String> {
    let data = ctx.data();
    let items_data = data.items_data.read();

    let mut types = std::collections::HashSet::new();
    for item in items_data.iter() {
        if !item.deleted && item.name != "None" {
            types.insert(item.main_type.clone());
        }
    }

    let mut type_vec: Vec<String> = types.into_iter().collect();
    type_vec.sort();

    type_vec
        .into_iter()
        .filter(|t| t.to_lowercase().contains(&partial.to_lowercase()))
        .take(25) // Discord limits to 25 choices
        .collect()
}

// Helper functions for sort command
pub fn build_sort_embed(
    sortable_items: &[SortableItem],
    stat_type: &str,
    item_type: Option<&str>,
    current_page: usize,
    total_pages: usize,
) -> CreateEmbed {
    let items_per_page = 10;
    let start_index = (current_page - 1) * items_per_page;
    let end_index = (start_index + items_per_page).min(sortable_items.len());

    let page_items = &sortable_items[start_index..end_index];

    let type_filter = get_type_filter(item_type);
    let stat_display = get_stat_display_name(stat_type);

    let mut description = format!(
        "Items sorted by {} at level 170{}\n\n",
        stat_display, type_filter
    );

    for (i, sortable_item) in page_items.iter().enumerate() {
        let rank = start_index + i + 1;
        let item = &sortable_item.item;

        description.push_str(&format!("**{}.** {}\n", rank, item.name));
        description.push_str(&format!(
            "   {}: **{}** | {}",
            stat_display, sortable_item.value, item.rarity
        ));

        if let Some(sub_type) = &item.sub_type {
            if !sub_type.is_empty() {
                description.push_str(&format!(" | {}", sub_type));
            }
        }

        description.push_str("\n\n");
    }

    CreateEmbed::new()
        .title(format!("Top Items by {}", stat_display))
        .description(description)
        .color(DEFAULT_COLOR)
        .footer(serenity::CreateEmbedFooter::new(format!(
            "Page {}/{} • {}",
            current_page, total_pages, EMBED_FOOTER
        )))
        .timestamp(serenity::Timestamp::now())
}

pub fn build_pagination_components(
    current_page: usize,
    total_pages: usize,
    stat_type: &str,
    item_type: Option<&str>,
) -> Vec<serenity::CreateActionRow> {
    let item_type_str = item_type.unwrap_or("");

    let mut prev_button = serenity::CreateButton::new(format!(
        "sort_prev_{}_{}_{}",
        stat_type, item_type_str, current_page
    ))
    .style(serenity::ButtonStyle::Secondary)
    .label("◀ Prev");

    if current_page <= 1 {
        prev_button = prev_button.disabled(true);
    }

    let page_button = serenity::CreateButton::new("sort_page_indicator")
        .style(serenity::ButtonStyle::Secondary)
        .label(format!("Page {}/{}", current_page, total_pages))
        .disabled(true);

    let mut next_button = serenity::CreateButton::new(format!(
        "sort_next_{}_{}_{}",
        stat_type, item_type_str, current_page
    ))
    .style(serenity::ButtonStyle::Secondary)
    .label("Next ▶");

    if current_page >= total_pages {
        next_button = next_button.disabled(true);
    }

    vec![serenity::CreateActionRow::Buttons(vec![
        prev_button,
        page_button,
        next_button,
    ])]
}
